src/Sarif/Writers/SarifConsolidator.cs

```diff
@@ -0,0 +1,259 @@
+        public bool RemoveWebRequests { get; set; }
+        public bool RemoveWebResponses { get; set; }
+
+        public SarifConsolidator(Run run)
+        {
+            _run = run;
+            _uniqueThreadFlows = new Dictionary<ThreadFlowLocation, int>(ThreadFlowLocation.ValueComparer);
+
+            if (_run.Artifacts != null)
+            {
+                // Realize Deferred Artifact list
+                _run.Artifacts = _run.Artifacts?.ToList();
+
+                // Remove indices from Run.Artifact array (redundant)
+                foreach (Artifact a in _run.Artifacts)
+                {
+                    if (a.Location != null)
+                    {
+                        a.Location.Index = -1;
+                    }
+                }
+            }
+
+            // Ensure Run TFL list is ready for consolidating TFLs
+            _run.ThreadFlowLocations = _run.ThreadFlowLocations ?? new List<ThreadFlowLocation>();
+
+            // Add existing TFLs to map (can't consolidate; there may be existing references to them)
+            for (int i = 0; i < _run.ThreadFlowLocations.Count; ++i)
+            {
+                ThreadFlowLocation tfl = _run.ThreadFlowLocations[i];
+                Trim(tfl.Location);
+                _uniqueThreadFlows[tfl] = i;
+            }
+        }
+
+        public void Trim(Result result)
+        {
+            if (result == null) { return; }
+
+            // Remove Result components if configured to
+            result.CodeFlows = (this.RemoveCodeFlows ? null : result.CodeFlows);
+            result.RelatedLocations = (this.RemoveRelatedLocations ? null : result.RelatedLocations);
+            result.Graphs = (this.RemoveGraphs ? null : result.Graphs);
+            result.GraphTraversals = (this.RemoveGraphTraversals ? null : result.GraphTraversals);
+            result.Stacks = (this.RemoveStacks ? null : result.Stacks);
+            result.WebRequest = (this.RemoveWebRequests ? null : result.WebRequest);
+            result.WebResponse = (this.RemoveWebResponses ? null : result.WebResponse);
+
+            // Truncate long Messages if configured
+            if (this.MessageLengthLimitBytes > 0)
+            {
+                if (result?.Message?.Text?.Length > this.MessageLengthLimitBytes)
+                {
+                    result.Message.Text = result.Message.Text.Substring(0, this.MessageLengthLimitBytes.Value);
+                }
+            }
+
+            // Trim Location and RelatedLocation properties
+            result.Locations = Trim(result.Locations);
+            result.RelatedLocations = Trim(result.RelatedLocations);
+
+            // Trim and Consolidate CodeFlows and ThreadFlows
+            ConsolidateCodeFlows(result);
+        }
+
+        public void ConsolidateCodeFlows(Result result)
+        {
+            if (result.CodeFlows != null)
+            {
+                foreach (CodeFlow cFlow in result.CodeFlows)
```

Comment:
>foreach [](start = 16, length = 7)

why isn't there a language feature for conditional enumeration over collections that can handle the case where the collection itself is null?

Reply:
I don't know. There isn't: 
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/foreach-in
https://stackoverflow.com/questions/6455311/is-ifitems-null-superfluous-before-foreacht-item-in-items

New trick from StackOverflow:
foreach(var item in collection ?? Enumerable<Type>.Empty) 

It's terse, but I don't love it.

Reply:
I'm going to leave as-is to better follow the design of the rest of the class. I generally null check each thing on the "Trim" entry of it, and check collections around both the foreach over elements and any collection-level manipulation together.